src/libmeasurement_kit/dns/system_resolver.hpp

```diff
@@ -0,0 +1,141 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include <measurement_kit/dns.hpp>
+
+#include <cassert>
+#include <future>
+
+namespace mk {
+namespace dns {
+
+class ResolverContext {
+  public:
+    QueryClass dns_class;
+    QueryType dns_type;
+    std::string name;
+    Callback<Error, Var<Message>> cb;
+    Settings settings;
+    Var<Reactor> reactor;
+    Var<Logger> logger;
+
+    addrinfo hints;
+    addrinfo *servinfo = nullptr;
+
+    Var<Message> message{new Message};
+
+    ResolverContext(QueryClass dns_c, QueryType dns_t, std::string n,
+                    Callback<Error, Var<Message>> c, Settings s, Var<Reactor> r,
+                    Var<Logger> l) {
+        dns_class = dns_c;
+        dns_type = dns_t;
+        name = n;
+        cb = c;
+        settings = s;
+        reactor = r;
+        logger = l;
+        memset(&hints, 0, sizeof(hints));
+    }
+
+    ~ResolverContext() {
+        if (servinfo != nullptr) {
+            freeaddrinfo(servinfo);
+        }
+    }
+};
+
+template<MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+static inline void resolve_async(ResolverContext *context) {
+    std::unique_ptr<ResolverContext> ctx(context);
+    Callback<Error, Var<Message>> callback = ctx->cb;
+    Var<Message> message_copy;
+
+    int error = getaddrinfo(ctx->name.c_str(), nullptr, &(ctx->hints),
+                            &(ctx->servinfo));
+    if (error) {
+        // check the error variable and return the correct error
+        ctx->logger->warn(gai_strerror(error));
+        ctx->reactor->call_soon(
+            [callback]() { callback(ResolverError(), nullptr); });
```

Comment:
I think we should discuss how to manage dns errors. Should I create new errors for the system_resolver or use the equivalent errors already used for the libevent resolver? I was planning to do it in another PR where I will integrate the system resolver with the library, add all the required tests and refactor the dns tests in:

    test/dns/query to test the dns functionalities
    test/dns/system_resolver to test the system_resolver function
    test/dns/libevent_resolver to test the libevent resolver